### PR TITLE
Trigger completion optimistically

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
I've tested it on `TypeChecker.fs`. No UI freezes at all (but completion appeared after ~2-3 seconds delay, however I expect it to be exactly 1 second, but it's another issue).

(no `Ctrl+Space` were used)

### CODE REVIEW

**Comment by saul** [comment]
Glad to see this revived and working well :) this is great.

**Comment by majocha** [comment]
Awesome, now it only needs more logic to filter out incorrect suggestions based on syntax, for example after `let`.

**Comment by vasily-kirichenko** [comment]
Yeah, it should not show anything right after let and let! However... What about `let (Case1 x) = ...`? It should show Case1.

**Comment by majocha** [comment]
Yes, it would be perfect to include only correct destructuring constructs. In C# it seems there is a lot of logic of this kind to make the experience fluid.

**Comment by cartermp** [comment]
@majocha That's correct, there is a lot of logic in there.

Completion for destructuring is tricky.  It would definitely be annoying if I had completion trigger for the following:

```fsharp
let (x
```

And my intention was to use pattern matching to form a tuple, with one of the names being `x`.  Perhaps constraining it only to upper-case?  I'm not sure.

**Comment by vasily-kirichenko** [comment]
@cartermp or you may want to write something like `let (module1.Module2.du1.Case1 x) = ...`

In short, in F# we can restrict completion list after dot only.

**Comment by cartermp** [comment]
Certainly some details to work out. I'll see if I can write anything useful down in #2432

**Comment by vasily-kirichenko** [comment]
@cartermp that thread is about _what_ to show in completion. The problem we are discussing in this PR is _when to auto trigger completion_. Those are completely different things. 

Currently we use simple parsing (was the last entered char a dot? is it a single dot? etc) and the lexer (are we inside a comment or excluded block? etc).

To provide good auto-triggering experience (i.e. non obtrusive), we'd have to use parse tree and a special worker, similar to the one used in filling the completion list (are we at argument position? are we in record construction? etc), but the logic would be different. 

So I think we should do not trigger completion anywhere, other then few very specific places in parse tree, which should be formalized. The first obvious place is on dot. **Please, suggest other places.**

**Comment by vasily-kirichenko** [comment]
@cartermp @Pilchie could you please help to find the code in C# editor which is responsible for triggering completion? All these http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CommonCompletionProvider.cs,54b4ce43ce7c9a05,references overrides do not have any real logic.

**Comment by vasily-kirichenko** [comment]
I've played with C# editor a bit. Completion triggers all the time. The only place that this PR's behavior is different is after `let` vs after `var` (C# do not trigger nor suggest anything at this point).

**Comment by majocha** [comment]
Another that comes to mind is after `type`. Generally in any place you may be declaring a new identifier completion should be up to the user.

**Comment by vasily-kirichenko** [comment]
@majocha Yeah. The problem here is that completion list is not empty at `type |` and `member __.Foo(|` (for instance) positions.

**Comment by vasily-kirichenko** [comment]
I turned off completion at the following points:
```fsharp
type |
type M|
let |
let f|
let f (|
let f(a|
let f(a, |
let f(a, b: int, |, c)
...
member __.M(|
member __.M(a, b|: int, c: int)
...
Please test for usability.

**Comment by cartermp** [comment]
@vasily-kirichenko Here's where I'd look: https://github.com/dotnet/roslyn/tree/master/src/Features/CSharp/Portable/Completion

I haven't pulled it down yet (not at my desktop), but my only other suggestion for *not* offering completion would be after the `.` in members:

```fsharp
member __.| // No completion
```

Also, I'm not sure if this is the PR for it, but we should also have preselection for locals that are in scope.

**Comment by nosami** [comment]
Some more cases that are accounted for in XS

```fsharp
let! |
override |
for |
[1..|
module |
(fun |
1u|
```

Not sure if you have these covered. XS doesn't handle record labels but it probably should. Anywhere where you can type an identifier IMO should not offer completions. Some tests [here](https://github.com/mono/monodevelop/blob/c92e333d91db093d193dd858e695806c17f5b2a7/main/external/fsharpbinding/MonoDevelop.FSharp.Tests/CompletionTests.fs#L57)

**Comment by vasily-kirichenko** [comment]
@nosami I think all these covered, except `let x =|`. Why do you not show completion in that case?

**Comment by vasily-kirichenko** [comment]
@cartermp it does not offer completion at the point you mention. I think it does not in master, too.

**Comment by nosami** [comment]
@vasily-kirichenko good question - no idea!

**Comment by vasily-kirichenko** [comment]
One of the failing tests expects completion list at `(*Mtypeparameter3*)` position:
```fsharp
namespace MyNamespace1
module MyModule =
    type DuType =
         | Tag of int          
    type Pet() = 
        member x.Name = "pet"
        member x.Speak() = "this is a pet"  
    type Dog() = 
        inherit Pet() 
namespace MyNamespace2
module MyModule2 =     
    let foo = MyNamespace1.MyModule(*Mtypeparameter1*)
    let f (x:int) = MyNamespace1.MyModule.DuType(*Mtypeparameter2*)    
    let typeFunc<MyNamespace1.MyModule(*Mtypeparameter3*)> = 10
```

I don't understand the `let typeFunc<some-concrete-type> = 10` syntax. Is it valid F# at all?

**Comment by dsyme** [comment]
> I don't understand the let typeFunc<some-concrete-type> = 10 syntax. Is it valid F# at all?

Yes, defines a type function, no explicit parameters but generic parameters inferred on use.  That's how ``typeof<..>`` and ``Set.empty`` works etc. e.g.

    let fail<'T> : 'T = failiwtih "fail"

    let f () = fail

Caveat emptor

**Comment by dsyme** [comment]
> I don't understand the let typeFunc<some-concrete-type> = 10 syntax. Is it valid F# at all?

No unless I'm missing something (my original answer was wrong)

**Comment by dsyme** [comment]
Very exciting PR - assuming it works well and stays performant it's a huge improvement to autocompletion in the Visual F# Tools

**Comment by Pilchie** [comment]
@rchande @cyrusnajmabadi - can you give some pointers on how C# does this?

**Comment by vasily-kirichenko** [comment]
@dsyme Thanks :) I tried to use this PR in real work (a bit)... Unfortunately, it's still too obtrusive :( Concretely, I don't like auto triggering when there is no meaningful parse tree at cursor point, which happens often (`like when you start to type `type` or `let` or `do` and completion appears). However, VS Code behaves even more aggressively, providing unrelated completion all the time, and people seem to like it... I don't know whether we should merge it without a flag in settings to disable it. As we've blocked by Roslyn in reusing its settings UI, I think we could add such a flag into our old-fashioned settings list (where VS FSI options are set) or, as a last resort, a environment variable, because people may go crazy by such annoying features.

**Comment by smoothdeveloper** [comment]
I'd not like the setting to be on and at the same time not be switchable.

We should merge it disabled and have an environment variable to switch it on for further testing, unless adding to the current F# settings is easy.

Completion in VS Code suffers from popping when the language service hasn't yet figured out the type and I find it annoying as I have to ESC, wait, and CTRL+Space frequently.

**Comment by vasily-kirichenko** [comment]
image

**Comment by smoothdeveloper** [comment]
It seems you found the right extension point for settings dialog :)

**Comment by vasily-kirichenko** [comment]
OK, I added a property page and a setting to turn auto triggering on and off. It's on by default. Should we make it disabled? If no, then this PR is ready.

**Comment by smoothdeveloper** [comment]
If we have setting, lets keep it on :) Thanks!

**Comment by vasily-kirichenko** [comment]
I turned off committing on space in auto triggering mode, so even though it pops up in some inappropriate places, it's not annoying because you don't have to press `Esc` all the time to dismiss suggestions. I find it very usable now.

File: vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs [hunk 1]
```diff
@@ -79,12 +150,16 @@ type internal FSharpCompletionProvider
...
-                shouldProvideCompletion(documentId, filePath, defines, sourceText, triggerPosition)
+                shouldProvideCompletion(documentId, filePath, defines, sourceText, triggerPosition) &&
+                (IntelliSenseSettings.ShowAfterCharIsTyped && 
+                 CommonCompletionUtilities.IsStartingNewWord(sourceText, triggerPosition, (fun ch -> isIdentifierStartCharacter ch), (fun ch -> isIdentifierPartCharacter ch)))
 
     static member ProvideCompletionsAsyncAux(checker: FSharpChecker, sourceText: SourceText, caretPosition: int, options: FSharpProjectOptions, filePath: string, textVersionHash: int) = 
         asyncMaybe {
             let! parseResults, parsedInput, checkFileResults = checker.ParseAndCheckDocument(filePath, textVersionHash, sourceText.ToString(), options, allowStaleResults = true)
 
+            //Logging.Logging.logInfof "AST:\n%+A" parsedInput
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: Commented code

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: Yes. It will be very useful in further work on completion.

**Comment by VSDekar** [comment]
In the moment we can use this in Visual Studio, you have recruited a new F# community member. I have never learned F# because it was so annoying when you have a C# Visual Studio Background and its also much easier to learn a new language with IntelliSense than without. So, thank you @vasily-kirichenko for your work!

**Comment by vasily-kirichenko** [comment]
@VSDekar Intellisense has worked since VS 2010, but it was not triggered automatically, except after dot. So saying that F# did not have intellisense is not fair.

**Comment by VSDekar** [comment]
@vasily-kirichenko You are right, but thats why i said that when you come from C# and Visual Studio. The experience is (was :-) ) much better with C#.

**Comment by smoothdeveloper** [comment]
@VSDekar I think each language has it's advantage, but confinning the overal experience to intellisense (which is working in F# since day 1) is short sighted, I also experience better error prone code constructs in C# and a greater need for heavy tooling to perform simple refactorings :smile: 

So you have to factor the whole experience with language and then draw a conclusion, but stopping at intellisense doesn't look as fancy as in other language is too early.

**Comment by KevinRansom** [comment]
@VSDekar 

Welcome to the team, we are very proud of the contributions of @vasily-kirichenko, @smoothdeveloper, @cloudRoutine  and our other community members they are truly allowing our tooling to catch up with the C# and VB experience.

Kevin

**Comment by smoothdeveloper** [comment]
@KevinRansom I'm not contributing to VS much, I'd need some help with #2351 if possible to try to help with IDE features at some point.

**Comment by pqnet** [comment]
The current (vs2017) trigger conditions are shit. Autocompletion triggers in places where syntax expects a new identifier (e.g. variable/type/member name/type parameter declaration) making it impossible to write a new identifier without pressing "esc" to close autocompletion.  Try to declare a type parameter, such as in `let myfun<'T> = ...` without the `'T` being autocompleted to `'tan`

**Comment by pqnet** [comment]
There is no reason for type name completion anywhere else than after a colon (`:` ) or `inherit` keyword. C# rules are a bad template for F# because while declarations in C# start with a type, declarations in F# start with name (and and optional type after colon)

**Comment by dsyme** [comment]
> Autocompletion triggers in places where syntax expects a new identifier (e.g. variable/type/member name/type parameter declaration) making it impossible to write a new identifier without pressing "esc" to close autocompletion. 


@pqnet Could you please open a new issue for that, with as many examples as you can? thanks

**Comment by cartermp** [comment]
Also @pqnet ensure you're on the latest VS (15.3.3) when you test, as there are changes there to not commit autocompletion except for a small number of cases, fixing some annoyances in this space.

In general please do be respectful here; autocompletion is difficult stuff and represents a lot of time and energy from @vasily-kirichenko and others.

**Comment by abelbraaksma** [comment]
>> Autocompletion triggers in places where syntax expects a new identifier (e.g. variable/type/member name/type parameter declaration) making it impossible to write a new identifier without pressing "esc" to close autocompletion.
>
> @pqnet Could you please open a new issue for that, with as many examples as you can? thanks

@dsyme, @vasily-kirichenko The same holds for trying to type an escaped identifier with ``` `` ``` (double backtick) syntax. I've reported this earlier here: #3406, but now you can't even type a single backtick-identifier that way without it will try to optimistically auto-complete after each space, comma, dot etc. The workaround is to constantly hit `Esc`. I'll update that issue accordingly.

Don't get me wrong, in any other code it usually works wonderfully and saves a lot of typing! But since 99% of my tests are written with this syntax, it makes writing them rather painful at the moment.

**Comment by pqnet** [comment]
@cartermp thank you, I was actually on 15.2 (didn't see any notification that a new version was available - I thought some would appear if my version were to be old) so probably my observation is obsolete. I'm sorry if my choice of words was a bit aggressive, I was frustrated and ended up being impolite. In my defense I tried to keep the criticism constructive and to give my contribution to the discussion: a concrete test case and my opinion about autocompletion triggering rules (there is no reason for it to trigger after every word, in particular after tokens which introduce a new name, like `let`, `for`, `<`, and similar).  @dsyme I'll check whether these issues or similar ones can be reproduced in the new version or not, and eventually file another issue ticket for that, as you suggested.

**Comment by CyrusNajmabadi** [comment]
The way C# handles this is with 'builders'.i.e. try writing:

```Func<int> f = s```

We will pop up completion because you may want to refer to something with 's' in it. However completion will be in 'builder' mode.  In this mode we do not commit by default. Instead the user has to explicitly pick an item and commit it  themselves. The builder message also says that we're in a special mode because you may be about to write a lambda. 

Basically the rules are:
If this is a reference-only location then open completion in normal mode.
If this is a location where you are only declaring a new variable name, do not pop up completion. 
If it is a location where both could be happening (like above) them year the builder.
